File: POTD/3309-count-prefix-and-suffix-pairs-i/count-prefix-and-suffix-pairs-i.cpp
class Solution {
public:
    int countPrefixSuffixPairs(vector<string>& words) {
        // Helper function to check if str1 is both a prefix and suffix of str2
        auto isPrefixAndSuffix = [](const string& str1, const string& str2) {
            return str2.find(str1) == 0 && str2.rfind(str1) == str2.size() - str1.size();
        };

        int count = 0;

        // Nested loops to check all pairs (i, j) where i < j
        for (int i = 0; i < words.size(); ++i) {
            for (int j = i + 1; j < words.size(); ++j) {
                if (isPrefixAndSuffix(words[i], words[j])) {
                    count++;
                }
            }
        }

        return count;
    }
};

File: POTD/3309-count-prefix-and-suffix-pairs-i/count-prefix-and-suffix-pairs-i.py
from typing import List

class Solution:
    def countPrefixSuffixPairs(self, words: List[str]) -> int:
        # Helper function to check if one string is both a prefix and suffix of another
        def isPrefixAndSuffix(str1: str, str2: str) -> bool:
            return str2.startswith(str1) and str2.endswith(str1)

        # Count the valid pairs (i, j) where i < j
        count = 0
        for i in range(len(words)):
            for j in range(i + 1, len(words)):
                if isPrefixAndSuffix(words[i], words[j]):
                    count += 1

        return count
